server/models/questions/getPublicContestQuestions.js

```diff
@@ -1,59 +1,133 @@
 const { pool } = require('../database')
 
+/**
+ *
+ * @param {Array} sortParams
+ * @return {String}
+ */
+function sortQuery(sortParams) {
+  if (!Array.isArray(sortParams)) {
+    return ''
+  }
+  return (
+    ' ORDER BY ' +
+    sortParams
+      .map(({ id, desc }) => `${id} ${desc ? 'DESC' : 'ASC'}`)
+      .join(', ')
+  )
+}
+
 /**
  * @param {*} param0
  * @param {Number} param0.cursor
  * @param {String} param0.search
  * @param {Number} param0.limit
  * @param {Number} param0.tag_ids
+ * @param {String} param0.difficulty
+ * @param {Number} param0.per_page_size
+ * @param {Number} param0.page
+ * @param {Array} param0.sort
  * @return {Promise}
  */
 
-function getPublicContestQuestions({ cursor, limit, search, tag_ids: tagIds }) {
+function getPublicContestQuestions({
+  search,
+  tag_ids: tagIds,
+  difficulty,
+  per_page_size: perPageSize,
+  page,
+  sort,
+}) {
   return new Promise((resolve, reject) => {
-    let query = `SELECT * FROM contests c
+    let countQuery = `SELECT COUNT(DISTINCT q.id) AS total FROM contests c
+                      INNER JOIN contests_questions cq ON c.id=cq.contest_id
+                      INNER JOIN questions q ON cq.question_id=q.id `
+    const cQArr = []
+
+    let query = `SELECT DISTINCT q.id, q.name, q.difficulty FROM contests c
                 INNER JOIN contests_questions cq ON c.id=cq.contest_id
                 INNER JOIN questions q ON cq.question_id=q.id `
     const qArr = []
 
-    if (tagIds && tagIds.length) {
+    if (
+      tagIds &&
+      Array.isArray(tagIds) &&
+      !tagIds.some(isNaN) &&
+      tagIds.length
+    ) {
+      countQuery += `INNER JOIN questions_tags qt ON q.id=qt.question_id WHERE qt.tag_id IN(`
       query += `INNER JOIN questions_tags qt ON q.id=qt.question_id WHERE qt.tag_id IN(`
+
       tagIds.forEach((tagId, idx) => {
         query += `?`
+        countQuery += `?`
         if (idx != tagIds.length - 1) {
           query += `,`
+          countQuery += `,`
         } else {
           query += `)`
+          countQuery += `)`
         }
         qArr.push(+tagId)
+        cQArr.push(+tagId)
       })
       query += ` AND `
+      countQuery += ` AND `
     } else {
       query += `WHERE `
+      countQuery += `WHERE `
     }
 
-    if (cursor) {
-      query += `q.id>? AND `
-      qArr.push(+cursor)
+    if (
+      difficulty &&
+      (difficulty === 'easy' ||
+        difficulty === 'medium' ||
+        difficulty === 'hard')
+    ) {
+      query += `difficulty=? AND `
+      countQuery += `difficulty=? AND `
+      qArr.push(difficulty)
+      cQArr.push(difficulty)
     }
 
     query += `c.end_time <= NOW() `
+    countQuery += `c.end_time <= NOW() `
```

Comment:
This endpoint is for practice questions and show return questions of only public contests and contests which do not contain confidential-questions.

```suggestion
    query += `c.end_time <= NOW()  AND c.public  = 1 AND c.confidential_questions = 0 `
    countQuery += `c.end_time <= NOW()  AND c.public = 1 AND c.confidential_questions = 0 `
```